core/src/main/java/io/parsingdata/metal/expression/value/reference/CurrentIteration.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2013-2018 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.expression.value.reference;
+
+import static io.parsingdata.metal.expression.value.ConstantFactory.createFromNumeric;
+
+import java.util.Optional;
+
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.data.ImmutableList;
+import io.parsingdata.metal.data.ParseState;
+import io.parsingdata.metal.encoding.Encoding;
+import io.parsingdata.metal.expression.value.Value;
+import io.parsingdata.metal.expression.value.ValueExpression;
+import io.parsingdata.metal.token.Rep;
+import io.parsingdata.metal.token.RepN;
+import io.parsingdata.metal.token.Token;
+
+/**
+ * A {@link ValueExpression} that represents the 0-based current iteration in an
+ * iterable {@link Token} (e.g. when inside a {@link Rep} or {@link RepN}).
+ */
+public class CurrentIteration implements ValueExpression {
+
+    @Override
+    public ImmutableList<Optional<Value>> eval(final ParseState parseState, final Encoding encoding) {
+        if (parseState.iterations.head == null) {
+            return ImmutableList.create(Optional.empty());
+        }
+        return ImmutableList.create(Optional.of(createFromNumeric(parseState.iterations.head, new Encoding())));
```

Comment:
The default `new Encoding()` is probably good enough here. For a minute I thought we needed to pass the `encoding` on, but I think this is correct.

Side note: Don't we have a `DEFAULT_ENCODING` available somewhere? I wonder how often we do `new Encoding()` like this.

Reply:
No, there is not a default encoding as constant available. We do this 6 times main; 4 in Shorthand, 1 in CurrentOffset, and 1 in CurrentIteration. In test scope we call it 7 times.

Reply:
Good point: #245